contracts/RewardManager.sol

```diff
@@ -319,13 +345,25 @@ contract RewardManager is Ownable, Versionable, Safe {
       msg.sender,
       rewardSafeOwner
     );
+
+    bool isAllowedGnosisExecution = _equalBytes(data, encodedTransactionData) &&
+      signatures[keccak256(contractSignature)] &&
+      isAllowedGnosisRecipient(to);
+
+    bytes4 result = isAllowedGnosisExecution ||
+      (to == address(this)) ||
+      eip1271Contracts.contains(to)
+      ? EIP1271_MAGIC_VALUE
+      : bytes4(0);
+
+    return result;
+  }
+
+  function isAllowedGnosisRecipient(address to) private view returns (bool) {
     return
-      ((_equalBytes(data, encodedTransactionData) &&
-        (to == msg.sender && signatures[keccak256(contractSignature)])) ||
-        (to == address(this)) ||
-        (eip1271Contracts.contains(to)))
-        ? EIP1271_MAGIC_VALUE
-        : bytes4(0);
+      (to == msg.sender) ||
+      TokenManager(ActionDispatcher(actionDispatcher).tokenManager())
+        .isValidToken(to);
```

Comment:
This is the implementation to see if it is an allowed token, is this check correct?

Reply:
Yup!

Reply:
Actually (sorry about this). This has me thinking on weather we want to have the `withdrawal` function on the contract. If we move the token to `eip1271Contracts` or another OR path, we don't even have to write a function on the reward manager. Meaning that a reward safe can literally do anything allowed on the token contract. We actually just have to move `signatures[keccak256(contractSignature)]` out of the way.

Reply:
But if we move `signatures[keccak256(contractSignature)]`  actually the issue with this is the protection from re-entry/replay attacks bcos I guess the principal is that we have to assume the token contracts are "untrusted".

Reply:
This might mean that `eip1271Contracts` pattern COULD be insecure. Right now it only crosses path with `claim` function which is the only real state-changing function, but I think the function abides by principals like state "changes first and external contract call last"